python_modules/dagster/dagster/_daemon/asset_daemon.py

```diff
@@ -572,309 +582,342 @@ def _process_auto_materialize_tick_generator(
-                    self._logger.warn(
-                        f"Moving dangling STARTED tick from evaluation {latest_tick.tick_data.auto_materialize_evaluation_id}{print_group_name} into SKIPPED"
+                # Evaluation ID will always be monotonically increasing, but will not always
+                # be auto-incrementing by 1 once there are multiple AMP evaluations happening in
+                # parallel
+                next_evaluation_id = self._get_next_evaluation_id()
+                tick = instance.create_tick(
+                    TickData(
+                        instigator_origin_id=instigator_origin_id,
+                        instigator_name=instigator_name,
+                        instigator_type=(
+                            InstigatorType.SENSOR if sensor else InstigatorType.AUTO_MATERIALIZE
+                        ),
+                        status=TickStatus.STARTED,
+                        timestamp=evaluation_time.timestamp(),
+                        selector_id=instigator_selector_id,
+                        auto_materialize_evaluation_id=next_evaluation_id,
                     )
-                    latest_tick = latest_tick.with_status(status=TickStatus.SKIPPED)
-                    instance.update_tick(latest_tick)
+                )
 
-        if retry_tick:
-            tick = retry_tick
-        else:
-            # Evaluation ID will always be monotonically increasing, but will not always
-            # be auto-incrementing by 1 once there are multiple AMP evaluations happening in
-            # parallel
-            next_evaluation_id = self._get_next_evaluation_id()
-            tick = instance.create_tick(
-                TickData(
-                    instigator_origin_id=instigator_origin_id,
-                    instigator_name=instigator_name,
-                    instigator_type=(
-                        InstigatorType.SENSOR if sensor else InstigatorType.AUTO_MATERIALIZE
-                    ),
-                    status=TickStatus.STARTED,
-                    timestamp=evaluation_time.timestamp(),
-                    selector_id=instigator_selector_id,
-                    auto_materialize_evaluation_id=next_evaluation_id,
+            with AutoMaterializeLaunchContext(
+                tick,
+                sensor,
+                instance,
+                self._logger,
+                tick_retention_settings,
+            ) as tick_context:
+                yield from self._evaluate_auto_materialize_tick(
+                    tick_context,
+                    tick,
+                    sensor,
+                    workspace_process_context,
+                    asset_graph,
+                    auto_materialize_asset_keys,
+                    stored_cursor,
+                    auto_observe_asset_keys,
+                    debug_crash_flags,
+                    sensor_state_lock,
+                    is_retry=(retry_tick is not None),
```

Comment:
this looks like a huge change, but is really just moving the logic that used to be here inside a separate _evaluate_auto_materialize_tick method that has a try/except around it